File: answers/leetcode/Flatten Binary Tree to Linked List/Flatten Binary Tree to Linked List.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        this.flattenAndGetTail(root);
    }

    private TreeNode flattenAndGetTail(TreeNode root) {
        if (null == root) {
            return null;
        }
        TreeNode leftHead = root.left;
        TreeNode leftTail = this.flattenAndGetTail(leftHead);
        TreeNode rightHead = root.right;
        TreeNode rightTail = this.flattenAndGetTail(rightHead);
        root.left = null;
        if (leftHead != null) {
            root.right = leftHead;
            leftTail.right = rightHead;
        } else {
            root.right = rightHead;
        }
        return rightTail != null ? rightTail : (leftTail != null ? leftTail : root);
    }
}

File: answers/leetcode/Flatten Binary Tree to Linked List/Flatten Binary Tree to Linked List.cpp
//@result 225 / 225 test cases passed. Status: Accepted Runtime: 8 ms Submitted: 0 minutes ago You are here! Your runtime beats 14.38% of cpp submissions.

// 2015-1-1
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	void flatten(TreeNode *root) {
		flattenRec(root);
	}
	TreeNode *flattenRec(TreeNode *root) {
		if (NULL == root) {
			return NULL;
		}
		TreeNode *left_end = flattenRec(root->left);
		if (left_end) {
			left_end->right = root->right;
			root->right = root->left;
			root->left = NULL;
		} else {
			left_end = root;
		}
		TreeNode *right_end = flattenRec(left_end->right);
		if (right_end) {
			return right_end;
		} else {
			return left_end;
		}
	}
};
